devices/ICub/src/ICub.cpp

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright (C) 2018 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#include "ICub.h"
+
+#include <yarp/os/LogStream.h>
+#include <yarp/os/Network.h>
+#include <yarp/os/BufferedPort.h>
+
+#include <assert.h>
+#include <map>
+#include <mutex>
+#include <string>
+#include <vector>
+
+using namespace wearable::devices;
+
+const std::string LogPrefix = "ICub :";
+
+class WrenchPort : public yarp::os::BufferedPort<yarp::os::Bottle> {
+public:
+    std::vector<double> wrenchValues;
+    bool firstRun = true;
+    mutable std::mutex mtx;
+
+    using yarp::os::BufferedPort<yarp::os::Bottle>::onRead;
+    virtual void onRead(yarp::os::Bottle& wrench) {
+        if (!wrench.isNull()) {
+            std::lock_guard<std::mutex> lock (mtx);
+
+            this->wrenchValues.clear(); // Clear the vector of previous values
+
+            this->wrenchValues.resize(wrench.size());
+            for (size_t i = 0; i < wrench.size(); ++i) {
+                this->wrenchValues.at(i) = wrench.get(i).asDouble();
+            }
+
+            if (this->firstRun)
+            {
+                this->firstRun = false;
+            }
+        }
+        else {
+            yWarning() << LogPrefix << "[WrenchPort] read an invalid wrench bottle";
+        }
+    }
+
+    std::vector<double> wrench() {
+        std::lock_guard<std::mutex> lck (mtx);
+        return this->wrenchValues;
+    }
+
+    bool firstRunFlag() {
+        return this->firstRun;
+    }
+};
+
+class ICub::ICubImpl
+{
+public:
+    class ICubForceTorque6DSensor;
+
+    wearable::TimeStamp timeStamp;
+
+    size_t nSensors;
+    std::vector<std::string> sensorNames;
+
+    bool leftHandFTDataReceived  = false;
+    bool rightHandFTDataReceived = false;
+
+    template<typename T>
+    struct ICubDeviceSensors
+    {
+        std::shared_ptr<T> icubSensor;
```

Comment:
At this point the entire struct is useless, use directly a `shared_ptr`.

Reply:
The struct provides a datatype based on the template which is used for creating a sensorMap https://github.com/Yeshasvitvs/wearables/blob/c58e258d7f8a3c2c6c7d22dfd3c80ce24cd04539/devices/ICub/src/ICub.cpp#L82

Reply:
Yes, is there any reason why you cannot have a `std::map<std::string, std::shared_ptr<ICubForceTorque6DSensor>>`?

Reply:
At this point it is not a problem to change it to specific sensor type but I prefer to use it as a template struct so that we can set a sensor map with different sensor types.

Reply:
Done